### ISSUE TITLE
UTM 4.6.x fails to boot into BSD ARM VMs

### ISSUE THREAD

**Issue by haihige** [utmapp/UTM#6918]
**Describe the issue**  
Starting with UTM 4.6.0, BSD ARM VMs fail to boot properly. I've run into this issue with FreeBSD and NetBSD. Linux doesn't seem to be affected.
The boot menu works (a prompt appears asking if I want to install the OS), but once it gets past that, it changes to "Display output is not active."
I tried both QEMU 9.1 and 7.2 (7.2 is used in 4.5.4), but the issue exists with both in 4.6.x. If I launch UTM 4.5.4, the BSD ARM VMs work correctly.
This can be tested with the standard ARM install images from FreeBSD and NetBSD, although it affects VMs that were created in previous version of UTM as well.

**Configuration**  
* UTM Version: 4.6.x
* macOS Version: 15.2
* Mac Chip (Intel, M1, ...): M3 Pro

**Debug log**  
[UTM 4.6.3 debug log](https://github.com/user-attachments/files/18231751/debug-4.6.3.log)
This debug log was grabbed from 4.6.3, but the issue persists in 4.6.4.

**Comment by fluff** [comment]
I have the same issue with all UTM 4.6.x on non-graphical Debian Linux 12 server VM installs. Works fine in 4.5.4, but all versions of UTM starting with 4.6.0, up to and including 4.6.4, shows "DIsplay output is not active" after the textmode GRUB bootloader starts loading the Linux kernel. (The textmode GRUB bootloader works correctly, but when the Linux kernel takes over, the display changes to "Display output is not active").

**Comment by naveenrajm7** [comment]
+1 
I tried installing multiple Linux Distros, and ran into this issue in some. 
I remember it happened in Debian, as soon as I start the installation the screen goes black. To avoid this I had to use Advanced Speech mode Install in Debian to avoid any GUI, another workaround was, trying to install using serial device rather than the display.

**Comment by TheRealKeto** [comment]
I don't experience this issue at all. 

Testing out on an M2 Macbook Air running Sonoma 14.7 with the VM image that FreeBSD distributes on the website for 14.2-RELEASE (`.qcow2`), the VM boots up, displays the "Display output is not active" message for 5 to 10 seconds, and then correctly displays the tty prompting a login. After rebooting, FreeBSD's boot menu correctly pops up without displaying the "Display output is not active" message.

As for existing VMs being affected, I also don't experience such issue. I had a VM set up for FreeBSD 14.2-RELEASE that was made using their official installer image on UTM 4.6.3, and recently just upgraded to UTM 4.6.4. The VM still safely boots.

**Comment by haihige** [comment]
> I don't experience this issue at all.
> 
> Testing out on an M2 Macbook Air running Sonoma 14.7 with the VM image that FreeBSD distributes on their website for 14.2-RELEASE (`.qcow2`), the VM boots up, displays the "Display output is not active" message for 5 to 10 seconds, and then correctly displays the tty prompting a login. After rebooting, FreeBSD's boot menu correctly pops up without displaying the "Display output is not active" message.
> 
> As for existing VMs being affected, I also don't experience such issue. I had a VM set up for FreeBSD 14.2-RELEASE that was made using their official installer image on UTM 4.6.3, and recently just upgraded to UTM 4.6.4. The VM still safely boots; this maybe an M3 issue?

Maybe this issue is just for VMs that boot the ISO, instead of using the VM image.

**Comment by TheRealKeto** [comment]
> Maybe this issue is just for VMs that boot the ISO, instead of using the VM image.
>  "...had a VM set up for FreeBSD 14.2-RELEASE that was made using their official installer image"

@haihige The FreeBSD's official installer is an ISO image. Both methods of making a VM work, at least on M2.

**Comment by haihige** [comment]
@TheRealKeto I see, I thought you meant the VM image (`.qcow2`) there, not the ISO.

**Comment by haihige** [comment]
Did some additional testing with three other Macs, all giving the same results (though slightly different than the M3):

1. M1 on macOS 14.5
2. M1 on macOS 15.1
3. M2 on macOS 15.2

I tested three ARM installer images:

1. NetBSD 10.0
2. FreeBSD 13.2
3. FreeBSD 14.1

With UTM 4.5.4, all three went past the boot menu fine and reached the installation screen. With 4.6.4, only FreeBSD 14.1 made it past the boot menu. FreeBSD 13.2 hangs on the "Display output is not active." screen and NetBSD 10.0 crashes `QEMULauncher` (crash log attached).

[QEMULauncher-2024-12-27-094434.ips.txt](https://github.com/user-attachments/files/18262613/QEMULauncher-2024-12-27-094434.ips.txt)

<img width="868" alt="Screenshot 2024-12-27 at 9 44 35 AM" src="https://github.com/user-attachments/assets/d345ff05-a35b-409d-915b-81a4200f066d" />

**Comment by fluxens** [comment]
Same issue with M2 pro on 15.2 (24C101) with Ubuntu 20.04 arm from UTMs gallery.

**Comment by gotenuoyd** [comment]
I am observing this on an M1 running macOS 14.7.3, UTM 4.6.4. I cannot boot NetBSD 10.0 (aarch64), it crashes QEMU with the same assertion on line 2065 that was mentioned earlier in the thread.

**Comment by MightySmallBalls** [comment]
I have the same issue on both installed and new installs on M4 Pro.

**Comment by kaznovac** [comment]
@haihige I've figured if I turn off the option `Use Hypervisor` option NetBSD no longer crashes the UTM, but after a few boot messages it gets stuck with `Display output is not active`

**Comment by manuelfhp** [comment]
I have the same issue trying to run FreeBSD 13.4 on MacBook Air M3, UTM 4.6.X
I can see the boot option but when FreeBSD starts to load comes a message "Display output is not Active"

I restore UTM old version 4.5.4 from time machine and now is running ok
If you want you can download from releases here in GitHub UTM page. Just browse the desired version in releases and click over the title. After that scroll down to the bottom to download the utm.dmg

I'll wait version UTM 4.7 to test if they fixed this issues.

**Comment by olegstepanov** [comment]
Have this problem when trying to install NetBSD 10.1 on MacBook Air M4 with UTM 4.7.4.

**Comment by manuelfhp** [comment]
Has this problem been fixed in the latest version 4.7.X ? Has anyone else experienced it?

Because I'm still using version 4.4.5 until it's finally fixed.

**Comment by haihige** [comment]
Just tested with FreeBSD 14.3-RELEASE using UTM 4.7.4 on an M4 Pro and it's working fine this time.

I'll close this for now, but anyone should feel free to reopen if they're still running in to this.

**Comment by manuelfhp** [comment]
But is not ok.
I can also use with FreeBSD 13.3-Release but in background mode, using a ssh  from terminal to connect in my VM.
But If you try to attach a Display and start using inside UTM to access using the console in UTM, when freebsd boots it will stop  and show a error message "Display Output is not Active"

![Image](https://github.com/user-attachments/assets/54d0b9d9-25de-436d-a9da-bbbcef4419d1)

@haihige are you able to access the display console inside UTM 4.7.4 using FreeBSD?

**Comment by haihige** [comment]
> [@haihige](https://github.com/haihige) are you able to access the display console inside UTM 4.7.4 using FreeBSD?

Yes, I was. I did not test with FreeBSD 13.3-RELEASE though, since that is [end of life](https://www.freebsd.org/security/#sup).

**Comment by manuelfhp** [comment]
Yes I can confirm.
I can run FreeBSD 14.3 in UTM 4.7.4
But FreeBSD 13.3 I can’t .

I don’t know why is not working under FreeBSD 13.X …the error message: "display output is not active”

Using UTM 4.5.4 I can run perfect.
Maybe some QEMU params or arguments is not compatible.

Is there any params or arguments that fix this problem?

Regards,
Manuel Pardo

**Comment by DesantBucie** [comment]
I have the same kind of problem with trying to boot qcow2 of 16-current or 15-relese. Either UEFI screen or Display output is not active, when I turn off UEFI boot.

**Comment by DesantBucie** [comment]
Neither does OpenBSD work, Linux boots just fine though.

**Comment by haihige** [comment]
Since other users can't reopen issues on this repo, so I'll reopen this.

**Comment by DesantBucie** [comment]
Well I'm sorry to say this but I might be just a giant dumbass. I didn't extract a .xz file and that's why I wasn't been able to boot FreeBSD. For OpenBSD I probably downloaded .img file instead of .iso. I did succesfully boot and install FreeBSD 15 RELEASE and lastest OpenBSD.

**Comment by kaznovac** [comment]
Stumbled on this
OpenBSD-current now runs as guest under Apple Hypervisor
https://www.undeadly.org/cgi?action=article;sid=20260115203619
> viogpu_wsmmap() returns a kva but instead should return a physical
address via [bus_dmamem_mmap(9)](https://man.openbsd.org/bus_dmamem_mmap). Without this, QEMU would only show a
black screen when starting X11. On the Apple Hypervisor, the kernel
would panic.

Might solve blank screen with apple hypervisor

**Comment by bigwaff** [comment]
I can confirm NetBSD 11.0 RC 1 will install from the official ISO image and boot successfully, but emulated display card needs to be set to 'virtio-gpu-pci' instead of default 'virtio-ramfb'. Also, Bridged Networking appears to assigned host IP address to VM using device 'virtio-net-pci', at least according to the output of ifconfig. 

UTM 4.7.5
MBP M1 Max
macOS Sequoia 15.7.3

I've only been successful with NetBSD 10.1 using UTM 4.5.4 where emulated display is 'virtio-ramfb'. Bridged Networking appears to assign host IP address to VM using device 'virtio-net-pci' according to ifconfig.
